File: 04.tax-simulator/java/src/test/java/tax/integration/TaxApiTests.java
package tax.integration;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.http.HttpStatus;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.within;

@SpringBootTest(classes = tax.simulator.TaxSimulatorApplication.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class TaxApiTests {
    private final RestTemplate restTemplate = new RestTemplate();

    @LocalServerPort
    private int port;

    @ParameterizedTest
    @CsvSource({
            "Célibataire, 2000, 0, 0, 1515.25",
            "Marié/Pacsé, 3000, 3000, 3, 3983.37"
    })
    void calculateTax_ReturnsOkStatusCode(String situationFamiliale, double salaireMensuel, double salaireMensuelConjoint, int nombreEnfants, double expectedResult) {
        var url = formatUrl(situationFamiliale, salaireMensuel, salaireMensuelConjoint, nombreEnfants);
        var response = restTemplate.getForObject(url, Double.class);

        assertThat(response).isCloseTo(expectedResult, within(0.01));
    }

    @ParameterizedTest
    @CsvSource({
            "Célibataire, -1000, 0, 0",
            "Célibataire, 3000, 0, -1"
    })
    void calculateTax_ReturnsBadRequest_ForInvalidInputs(String situationFamiliale, double salaireMensuel, double salaireMensuelConjoint, int nombreEnfants) {
        var url = formatUrl(situationFamiliale, salaireMensuel, salaireMensuelConjoint, nombreEnfants);
        try {
            restTemplate.getForEntity(url, String.class);
        } catch (HttpClientErrorException e) {
            assertThat(e.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
        }
    }

    private String getBaseUrl() {
        return "http://localhost:" + port + "/api/tax/calculate";
    }

    private String formatUrl(String situationFamiliale, double salaireMensuel, double salaireMensuelConjoint, int nombreEnfants) {
        return String.format("%s?situationFamiliale=%s&salaireMensuel=%f&salaireMensuelConjoint=%f&nombreEnfants=%d",
                getBaseUrl(), situationFamiliale, salaireMensuel, salaireMensuelConjoint, nombreEnfants);
    }
}

File: 03.tax-simulator/csharp/Tax.Simulator.Tests/Integration/TaxApiTests.cs
using System.Net;
using System.Net.Http.Json;
using FluentAssertions;
using Microsoft.AspNetCore.Mvc.Testing;
using Xunit;

namespace Tax.Simulator.Tests.Integration;

public class TaxApiTests(WebApplicationFactory<Program> factory) : IClassFixture<WebApplicationFactory<Program>>
{
    private readonly HttpClient _client = factory.CreateClient();

    [Theory]
    [InlineData("Célibataire", 2000, 0, 0, 1515.25)]
    [InlineData("Marié/Pacsé", 3000, 3000, 3, 3983.37)]
    public async Task CalculateTax_ReturnsOkStatusCode(
        string situationFamiliale,
        decimal salaireMensuel,
        decimal salaireMensuelConjoint,
        int nombreEnfants,
        decimal expectedResult)
    {
        var response =
            await _client.GetAsync(FormatUrl(situationFamiliale, salaireMensuel, salaireMensuelConjoint,
                nombreEnfants));
        response.StatusCode.Should().Be(HttpStatusCode.OK);

        var result = await response.Content.ReadFromJsonAsync<decimal>();
        result.Should().Be(expectedResult);
    }

    [Theory]
    [InlineData("Célibataire", -1000, 0, 0)]
    [InlineData("Célibataire", 3000, 0, -1)]
    public async Task CalculateTax_ReturnsBadRequest_For_Invalid_Inputs(
        string situationFamiliale,
        decimal salaireMensuel,
        decimal salaireMensuelConjoint,
        int nombreEnfants)
        => (await _client.GetAsync(
                FormatUrl(situationFamiliale, salaireMensuel, salaireMensuelConjoint, nombreEnfants))
            ).StatusCode
            .Should()
            .Be(HttpStatusCode.BadRequest);

    private static string FormatUrl(
        string situationFamiliale,
        decimal salaireMensuel,
        decimal salaireMensuelConjoint,
        int nombreEnfants) =>
        $"/api/tax/calculate?situationFamiliale={situationFamiliale}&salaireMensuel={salaireMensuel}&salaireMensuelConjoint={salaireMensuelConjoint}&nombreEnfants={nombreEnfants}";
}